### PR TITLE
 Consolidate all calls to unidecode & remove unidecode from form field clean_name

### PR DESCRIPTION
**Pull request by lb-**
This PR covers a few items that are interrelated, I am happy to split into two PRs if that is more suitable.

The underlying goal is to centralise the usage of `unidecode` in preparation for a future removal of reliance on this package due to licensing issues - see #3311

## Overview
* *Builds on the work by @ashiazed & @mvantellingen in separate PRs*
* Moves `unidecode` to a central wagtail.core.utils function so it can be replaced/removed easier in the future, superseeds the work done on #5508
* Provides a safe migration path for existing applications with Wagtail Forms where saved data must still honour the `unidecode` kebab-case clean name on fields.
* Wagtail forms will now generate a clean name on fields as a field, not a method, this provides the ability to do the migration path above but has been done in a way that no longer relies on `unidecode`. As per https://github.com/wagtail/wagtail/pull/3088#issuecomment-284877264
* Finally, in the change for the clean name on form fields, we have adopted `snake_case` instead of `kebab-case`, this makes these easier to use in templates (for simple cases), superseeds #3088
* Small fixes; add some basic tests for the camel case conversion, add tests for the new wrapper function around `unidecode`, refine duplicate field labels to compare on their eventual clean_name instead of just a basic string comparison.

## How to Validate this PR
* First, ensure you have an app running locally that has at least one form page, ideally two.* Add some new fields to each form, try to use some unicode values inside the labels - https://www.ltg.ed.ac.uk/~richard/unicode-sample.html
* Submit some responses on each of those forms
* Check out the PR on the local app, first run manage.py runserver, you should receive a warning that there are migrations to run AND the application should not crash.
* Run makemigrations, the existing form field model should be updated with an addied field `clean_name`, then migrate. When running these, it should not crash.
* Finally, load up the server, this is where the checks should run and because the fields are added you should get an info message advising that some clean_name updates have happened.* Check that you can still see the entries saved as form responses previously.
* Attempt to create a duplicate field in an existing page with some caps/non-caps variation of an existing field.

## Notes
* The new clean names may not be as visually nice for complex unicode scenarios, however these are used internally only and only those doing customisations in rendering will get visibility into these, even when exporting to csv the actual user entered label is resolved.
* We can refine the `safe_snake_case` function in the future knowing it will not break existing data stored.
* This PR does not handle scenarios where a field is added to a form with name 'A' and then changed to name 'B' and updating the clean_name from there, this is intentional so that data is not lost when labels change. Plus implementing this would open us up to a lot of complexity.
* There are massive changes to the form tests and test data, however I have intentionally set one model in the test data to have clean_name blank. This is to ensure we can test that the checks run and will update data with the legacy clean name.
* This PR does not remove the requirement of unidecode, but paves the way for this to be done in the next release as there is now a clear migration strategy for existing applications.

## Checks
* Do the tests still pass? 👍 
* Does the code comply with the style guide? 👍 
* For Python changes: Have you added tests to cover the new/fixed behaviour? 👍 
* For new features: Has the documentation been updated accordingly? N/A - we may want to add some upgrade considerations but the intent of this change is that it happens seamlessly.

### CODE REVIEW

**Comment by lb-** [comment]
Getting some test db migration errors, will investigate tomorrow. 

```
===================

FAIL: test__migrations (wagtail.core.tests.test_migrations.TestForMigrations)

----------------------------------------------------------------------

Traceback (most recent call last):

File "/home/travis/build/wagtail/wagtail/wagtail/core/tests/test_migrations.py", line 54, in test__migrations

self.fail('Model changes with no migrations detected:\n%s' % migrations)

AssertionError: Model changes with no migrations detected:

tests.0054_auto_20200531_1914

Alter field name on restauranttag

Alter field slug on restauranttag


```

**Comment by gasman** [comment]
@lb- That looks like the migration change in Taggit fixed by #6071 / #6075.

**Comment by lb-** [comment]
@gasman  - thanks, you were right.

File: wagtail/contrib/forms/models.py [hunk 1]
```diff
@@ -115,6 +119,61 @@ def clean_name(self):
...
+        """
+        Ensure that existing data stored will be accessible via the legacy clean_name.
+        When checks run, replace any blank clean_name values with the unidecode conversion.
+        """
+
+        try:
+            objects = cls.objects.filter(clean_name__exact='')
+
+            for obj in objects:
+                legacy_clean_name = str(slugify(string_to_ascii(obj.label)))
```
Comment (gasman) [inline wagtail/contrib/forms/models.py:149]: I'm concerned about the use of `string_to_ascii` here, as we're not distinguishing between uses that rely on it preserving unidecode semantics (i.e. this one) and ones that just want an ASCII string and don't care how (i.e. the HTTP headers for document serving) - I think it'd better to use `unidecode` directly here. If we do it as follows, I think we can avoid having to revisit this code later when we drop unidecode:

* if `cls.objects.filter(...)` returns no results, return immediately
* Otherwise, import unidecode in a try/catch
* If that throws an ImportError, rethrow it with an explanatory message, something like: "You have form submission data that was created on an older version of Wagtail and requires the unidecode library to retrieve it correctly. Please install the unidecode package." (Currently this should never happen on a properly installed Wagtail instance, while unidecode is listed as a dependency in setup.py, but this will come into play when we drop that dependency, for anyone who upgrades directly from 2.9 to that version.)
* Otherwise, proceed to populate clean_name, using `unidecode` to convert the label

At that point, I think it would be reasonable to drop `string_to_ascii` entirely and use `safe_snake_case` in the remaining spots instead. Then, the only thing left to do later (in 2.12, if we follow the usual deprecation timeline) is to drop the unidecode dependency from setup.py.

File: wagtail/contrib/forms/utils.py [hunk 2]
```diff
@@ -2,10 +2,19 @@
 
 from wagtail.core import hooks
 from wagtail.core.models import UserPagePermissionsProxy, get_page_models
+from wagtail.core.utils import safe_snake_case
 
 _FORM_CONTENT_TYPES = None
 
 
+def get_field_clean_name(label, legacy=False):
```
Comment (gasman) [inline wagtail/contrib/forms/utils.py:10]: The `legacy` kwarg is unused here and can be removed.

**Comment by lb-** [comment]
Thanks for the feedback Matt, will get on to this soon.

**Comment by lb-** [comment]
updated with requested changes (hopefully I understood correctly), also rebased to master + updated the migration file to suit latest master.

**Comment by lb-** [comment]
Fixed linting error (force pushed)

**Comment by gasman** [comment]
Thanks @lb-! Merged in 0546237a1526cc2d17bb37f4503583ffd6b62715 + parents - I think we can reasonably take the next step of swapping out unidecode from `string_to_ascii`, but now that we've got the deprecation path in place there's no rush to get that follow-up change into 2.10 too - especially if we're going to evaluate [anyascii](https://pypi.org/project/anyascii/) as a replacement.

File: wagtail/contrib/forms/models.py [hunk 3]
```diff
@@ -115,6 +119,61 @@ def clean_name(self):
...
+                obj.clean_name = legacy_clean_name
+                obj.save()
+
+            if objects.count() > 0:
+                return Info(
+                    'Added `clean_name` on %s form field(s)' % objects.count(),
+                    obj=cls
+                )
+
+        except FieldError:
```
[the remark this answers is not in the document]
Comment (gasman) [reply wagtail/contrib/forms/models.py:147]: Oops, looks like that was bad advice - on an unmigrated bakerydemo, I'm now getting

```
django.db.utils.ProgrammingError: column base_formfield.clean_name does not exist
LINE 1: ...COUNT(*) AS "__count" FROM "base_formfield" WHERE "base_form...
```
presumably because the database query (and thus the error) only happens on the `objects.count()` line below. Will fix...

[the remark this answers is not in the document]
Comment (gasman) [reply wagtail/contrib/forms/models.py:147]: Fixed in b8c36c0d7df23f19f3be61e28246f32484c26644

**Comment by lb-** [comment]
Awesome! Thanks @gasman super cool to see this in.
